### ISSUE TITLE
file.line ensure before behavior doesn't handle multiple matches (multiple blocks) well

### ISSUE THREAD

**Issue by shallot** [saltstack/salt#40821]
It would be nice if this worked:

```
epoll enable:
  file.line:
    - name: /etc/nginx/nginx.conf
    - mode: ensure
    - after: 'events {'
    - content: '        use epoll;'
    - indent: True
    - before: '}'
```

instead of throwing:

```
----------
          ID: epoll enable
    Function: file.line
        Name: /etc/nginx/nginx.conf
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/.salt-ssh-thin-dir/py2/salt/state.py", line 1746, in call
                  **cdata['kwargs'])
                File "/.salt-ssh-thin-dir/py2/salt/loader.py", line 1703, in wrapper
                  return f(*args, **kwargs)
                File "/.salt-ssh-thin-dir/py2/salt/states/file.py", line 3191, in line
                  backup=backup, quiet=quiet, indent=indent)
                File "/.salt-ssh-thin-dir/py2/salt/modules/file.py", line 1746, in line
                  _assert_occurrence(body, before, 'before')
                File "/.salt-ssh-thin-dir/py2/salt/modules/file.py", line 1534, in _assert_occurrence
                  raise CommandExecutionError('Found {0} expected occurrences in "{1}" expression'.format(msg, target))
              CommandExecutionError: Found more than expected occurrences in "before" expression
     Started: 15:06:29.952029
    Duration: 1.767 ms
     Changes:
```

This happens because the file has multiple blocks of:

```
foo {
  ...
}
bar {
  ...
}
baz {
  ...
}
```

Yet, the annoying part is that empirical testing shows that `after` and `before` may only be a single line apart. So why is the parser continuing on through the entire file if the layout is effectively:

```
match after 1
non-match 1
match before 1
non-match 2
match before 2
non-match 3
match before 3
```

The only one that could possibly work with `match after 1` is `non-match 1` -- the existence of `match before 2` and `match before 3` is supposed to be inconsequential.

I noticed also that there's a line in that code saying:

```
                raise CommandExecutionError('Found more than one line between boundaries "before" and "after".')
```

Unfortunately, I never actually had this appear in practice. It would have been very useful if it had!

### Versions Report

```
Salt Version:
           Salt: 2016.11.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Oct 26 2016, 20:30:19)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: Not Installed
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: Not Installed

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-108-generic
         system: Linux
        version: Ubuntu 14.04 trusty
```

**Comment by gtmanfred** [comment]
@isbm would you be able to help me with this?

From looking at the code it looks like only one of `before` or `after` is really supposed to be used?

And that is what is causing this error.

Thanks,
Daniel

**Comment by isbm** [comment]
Yes, either one.

**Comment by shallot** [comment]
FWIW there does appear to be code explicitly handling "if before and after". And it makes some logical sense to have such an option - if it enables reasonable block editing. For example, if the user wants to edit a file containing:

```
<Directory /foo>
  Something default you don't want to lose
  Logical place to insert new stuff
</Directory>

<Directory /bar>
  Something default you don't want to lose
  Logical place to insert new stuff
</Directory>
```

When you want to ensure only the `/foo` block has something, and put it into ```Logical place to insert new stuff```, it would be nice to have the option to match specifically ```Directory /bar>``` as well as the next ```</Directory>``` immediately after it in order to be able to keep things in the right order.

Arguably the need for these kinds of things tends to diminish with time, as more and more people rewrite their config files to utilize more 'include' features, but it does still exist in the wild. Heck, even if you make it a priority not to do any more controversial single-line edits, it still stands to reason that you may need to do one single-line edit - to make sure that the file utilizes an 'include' feature.

**Comment by gtmanfred** [comment]
Yup, i completely agree that it should exist, it just doesn't appear to.

I am going to mark this issue as a Feature request to add that in the
future.

If you have time and are inclined, we would very much appreciate a Pull
request implementing that.

Thanks,
Daniel

**Comment by shellhacks** [comment]
Since this feature is yet not released ... does anybody know  how to perform such kind of insert? Maybe any workaround.
